Generic overload resolution fails in for-in statement

**Describe the bug**
When using a `Collection` method with multiple overloads in `for`-`in` loop inside a generic context, the compiler reports ambiguity where there is none.

**To Reproduce**
Attempt to compile the following code ([also attached](https://github.com/apple/swift/files/8925239/test.swift.txt)):

```swift
func f<C: Collection>(_ c: C) {
  for _ in c.prefix(1) {}
}
```

Compilation fails with this error message:

```
test.swift:3:12: error: ambiguous use of 'prefix'
  for _ in c.prefix(1) {}
           ^
Swift.Collection:6:40: note: found this candidate
    @inlinable public __consuming func prefix(_ maxLength: Int) -> Self.SubSequence
                                       ^
Swift.Sequence:7:40: note: found this candidate
    @inlinable public __consuming func prefix(_ maxLength: Int) -> PrefixSequence<Self>
                                       ^
```

**Expected behavior**
Successful compilation.

**Environment**
The earliest snapshot that fails to compile is from 2022-06-07:
```
Apple Swift version 5.8-dev (LLVM fdfb8ac496e4538, Swift c9f10040f7f3838)
```
The code above compiles correctly using the 2022-06-02 snapshot:
```
Apple Swift version 5.8-dev (LLVM 278d67f38c6a910, Swift ee312bc1e20eb01)
```

**Additional context**
Assigning the result of `c.prefix(1)` before the loop or using it in a `for`-`in` loop in a non-generic context both succeed.

```swift
// OK
func g<C: Collection>(_ c: C) {
  let pre = c.prefix(1)
  for _ in pre {}
}

// OK
func h(_ c: Array<Int>) {
  for _ in c.prefix(1) {}
}
```

--- a/lib/Sema/CSGen.cpp
+++ b/lib/Sema/CSGen.cpp
@@ -3944,8 +3944,20 @@ generateForEachStmtConstraints(
     forEachStmtInfo.makeIteratorVar = PB;
 
     // Type of sequence expression has to conform to Sequence protocol.
+    //
+    // Note that the following emulates having `$generator` separately
+    // type-checked by introducing a `TVO_PrefersSubtypeBinding` type
+    // variable that would make sure that result of `.makeIterator` would
+    // get ranked standalone.
     {
-      cs.addConstraint(ConstraintKind::ConformsTo, cs.getType(sequenceExpr),
+      auto *externalIteratorType = cs.createTypeVariable(
+          cs.getConstraintLocator(sequenceExpr), TVO_PrefersSubtypeBinding);
+
+      cs.addConstraint(ConstraintKind::Equal, externalIteratorType,
+                       cs.getType(sequenceExpr),
+                       externalIteratorType->getImpl().getLocator());
+
+      cs.addConstraint(ConstraintKind::ConformsTo, externalIteratorType,
                        sequenceProto->getDeclaredInterfaceType(),
                        contextualLocator);
 

--- a/test/stmt/foreach.swift
+++ b/test/stmt/foreach.swift
@@ -245,3 +245,8 @@ func testForEachWhereWithClosure(_ x: [Int]) {
   for i in x where x.contains(where: { $0.byteSwapped == i }) {}
 }
 
+// https://github.com/apple/swift/issues/59522 - use of `prefix` with generic base causes ambiguity in for-in statement
+func test_no_ambiguity_with_prefix_iterator<C: Collection>(c: C) {
+  for _ in c.prefix(1) { // Ok
+  }
+}

